cmd/pdatagen/internal/base_slices.go

```diff
@@ -30,6 +30,26 @@ func (es ${structName}) MoveAndAppendTo(dest ${structName}) {
 		*dest.orig = append(*dest.orig, *es.orig...)
 	}
 	*es.orig = nil
+}
+
+// RemoveIf calls f sequentially for each element present in the slice.
+// If f returns true, the element is removed from the slice.
+func (es ${structName}) RemoveIf(f func(${elementName}) bool) {
```

Comment:
What do you think about a slightly different API, a `RemoveAt(index int)`? The caller can iterate using Len/At and search for the element they want to find, then call RemoteAt. I think it should have the same performance and is simpler to understand.

It would be interesting to compare the performance. IIRC callback func are pretty slow in Go, so my proposed variation may even be faster if Len and At are properly inlined.

Reply:
Indeed calling f was proven to be a bit slower than the At way, but removing an element at position ix and keeping order is O(N) then you get O(n2) for your proposal, this proposal is doing everything in O(N), so  I am confident it cannot be faster if we preserve order.

Reply:
I am suggesting this:
```go
for i:=0; i<slice.Len(); i++ {
  elem := slice.At(i)
  if someCond(elem) {
    slice.RemoveAt(i)
    break
  }
}
```
This is O(n) and preserves the order.

Reply:
But slice.RemoveAt(I) is O(n) unless you do the swap with last one which does not preserve the order

Reply:
Correct, but it is the same amount of work that `RemoveIf` does, right?

Reply:
Here are some benchmarks:
```
goos: darwin
goarch: amd64
pkg: go.opentelemetry.io/collector/consumer/pdata
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkMetrics_RemoveIf
BenchmarkMetrics_RemoveIf-16    	 1547742	       764.1 ns/op	    1024 B/op	       1 allocs/op
PASS

goos: darwin
goarch: amd64
pkg: go.opentelemetry.io/collector/consumer/pdata
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkMetrics_RemoveAt
BenchmarkMetrics_RemoveAt-16    	  987662	      1023 ns/op	    1024 B/op	       1 allocs/op
PASS
```

And the code looks a bit harder to write because you need to ensure that you go back with the index:

```golang
func BenchmarkMetrics_RemoveIf(b *testing.B) {
	req := make([]*otlpmetrics.Metric, 126)
	for i := 0; i < 126; i++ {
		req = append(req, generateTestProtoIntGaugeMetric())
	}

	b.ReportAllocs()
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		cloneReq := make([]*otlpmetrics.Metric, 126)
		copy(cloneReq, req)
		slice := newMetricSlice(&cloneReq)
		index := 0
		slice.RemoveIf(func(metric Metric) bool {
			index++
			return index%3 == 0
		})
		if slice.Len() != 84 {
			b.Fail()
		}
	}
}

func BenchmarkMetrics_RemoveAt(b *testing.B) {
	req := make([]*otlpmetrics.Metric, 126)
	for i := 0; i < 126; i++ {
		req = append(req, generateTestProtoIntGaugeMetric())
	}

	b.ReportAllocs()
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		cloneReq := make([]*otlpmetrics.Metric, 126)
		copy(cloneReq, req)
		slice := newMetricSlice(&cloneReq)
		index := 0
		for i := 0; i < slice.Len(); {
			index++
			if index%3 == 0 {
				copy((*slice.orig)[i:], (*slice.orig)[i+1:])
				*slice.orig = (*slice.orig)[:len(*slice.orig)-1]
				continue
			}
			i++  // Move i only when we don't remove
		}
		if slice.Len() != 84 {
			b.Fail()
		}
	}
}
```

Reply:
Yeah, I agree, you need to be slightly more careful with RemoveAt if you want to remove them all.

Reply:
With RemoveIf you're only constructing a new slice after elements have all been removed. With RemoveAt you have a new slice created each time you call. Perhaps that's where some additional overhead is coming from?

Reply:
I don't know why we need to construct a new slice. It should be an inplace operation.